### PR TITLE
[SVS-9] Show info bar when the quality profile has changed

### PR DESCRIPTION
**Pull request by eyalantzMS**
1. Will kick off a background task that will check if the quality profile has changed
2. Will still work on older SQ servers (<5.2) but will not perform the timestamp check since there's nothing to check.
3. If the solution closed/reopened it will call a refresh which will cancel the background task
4. Once the new sqconfig is used, opening a solution with it in a different machine (with previous version of the extension) will work as well, obviously without any update suggestions.

Misc: renamed RuleSetGroup -> LanguageGroup since its a bit more general than what it was when we started. I also moved various related helpers into a separate common class.

### CODE REVIEW

File: src/Integration.UnitTests/Binding/BindingWorkflowTests.cs [hunk 1]
```diff
@@ -81,14 +81,15 @@ public void BindingWorkflow_DownloadQualityProfile_Success()
             var additionalFiles = new[] { new AdditionalFile { FileName = "abc.xml", Content = new byte[] { 1, 2, 3 } } };
             RoslynExportProfile export = RoslynExportProfileHelper.CreateExport(ruleSet, nugetPackages, additionalFiles);
 
-            var language = Language.CSharp;
-            this.ConfigureProfileExport(testSubject, export, language, RuleSetGroup.VB);
+            var language = Language.VBNET;
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Binding/BindingWorkflowTests.cs]: No longer support injecting other mappings.

File: src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs [hunk 2]
```diff
@@ -0,0 +1,386 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs" company="SonarSource SA and Microsoft Corporation">
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs]: Separate test class for ErrorListInfoBarController.QualityProfileBackgroundProcessor.
I also added a test at the ErrorListInfoBarController level that checks for cancellation and have another that will also invoke the background processor and will no-op before executing the task.

File: src/Integration.UnitTests/Persistence/SolutionBindingSerializerTests.cs [hunk 3]
```diff
@@ -102,6 +103,46 @@ public void SolutionBindingSerializer_WriteSolutionBinding_ReadSolutionBinding()
         }
 
         [TestMethod]
+        public void SolutionBindingSerializer_WriteSolutionBinding_ReadSolutionBinding_WithProfiles()
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Persistence/SolutionBindingSerializerTests.cs]: Just to cover the profiles

File: src/Integration/Binding/BindingWorkflow.cs [hunk 4]
```diff
@@ -34,12 +34,6 @@ internal class BindingWorkflow
         private readonly IProjectSystemHelper projectSystem;
         private readonly SolutionBindingOperation solutionBindingOperation;
 
-        internal readonly Dictionary<Language, RuleSetGroup> LanguageToGroupMapping = new Dictionary<Language, RuleSetGroup>
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Became a method in LanguageGroupHelper class

File: src/Integration/Service/SonarQubeServiceWrapper.cs [hunk 5]
```diff
@@ -222,21 +262,10 @@ public Uri CreateProjectDashboardUrl(ConnectionInformation connectionInformation
 
         internal /*for testing purposes*/ static string CreateQualityProfileExportUrl(QualityProfile profile, string language, string exporter)
         {
+            // TODO: why name and not key? why language is needed at all, profiles are per language
```
Comment (eyalantzMS) [inline src/Integration/Service/SonarQubeServiceWrapper.cs]: Sent an email about it, Matthew was copied, in case we'll get a reply when I'm not around and we need to do something.

File: src/SonarQube/SolutionBinding.sqconfig [hunk 6]
```diff
@@ -1,4 +1,10 @@
 {
   "ServerUri": "https://dory.sonarsource.com/sonarqube/",
-  "ProjectKey": "sonarlint-visualstudio"
+  "ProjectKey": "sonarlint-visualstudio",
+  "Profiles": {
```
Comment (eyalantzMS) [inline src/SonarQube/SolutionBinding.sqconfig]: This is how the file will look now

File: src/Integration/Persistence/BoundSonarQubeProject.cs [hunk 7]
```diff
@@ -38,6 +39,8 @@ public BoundSonarQubeProject(Uri serverUri, string projectKey, ICredentials cred
 
         public string ProjectKey { get; set; }
 
+        public Dictionary<LanguageGroup, ApplicableQualityProfile> Profiles { get; set; }
```
Comment (eyalantzMS) [inline src/Integration/Persistence/BoundSonarQubeProject.cs]: Decided on a project
1) Reads nice as json (see example in this PR)
2) We need to compare a profile for language vs. a profile for the same language, dictionary provides that aggregation by definition.

File: src/Integration/Binding/BindingWorkflow.cs [hunk 8]
```diff
@@ -207,8 +206,17 @@ private ProgressStepDefinition[] CreateWorkflowSteps(IProgressController control
             foreach (var language in languages)
             {
                 notifier.NotifyCurrentProgress(string.Format(CultureInfo.CurrentCulture, Strings.DownloadingQualityProfileProgressMessage, language.Name));
+
+                QualityProfile qualityProfileInfo;
+                if (!host.SonarQubeService.TryGetQualityProfile(this.connectionInformation, this.project, language.ServerKey, cancellationToken, out qualityProfileInfo))
+                {
+                    failed = true;
+                    break;
```
Comment (duncanpMS) [inline src/Integration/Binding/BindingWorkflow.cs]: What will _TryGetQualityProfile_ return if the particular language plugin isn't installed on the server? (there's no guarantee that either the C# or VB plugins will be installed). If the call will return false in that case then breaking out of the loop and not checking the other languages probably isn't the right thing to do.

File: src/Integration.UnitTests/LanguageGroupHelperTests.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
+{
+    [TestClass]
+    public class LanguageGroupHelperTests
+    {
+        [TestMethod]
+        public void LanguageGroupHelper_GerLanguage()
+        {
+            Assert.AreSame(Language.CSharp, LanguageGroupHelper.GerLanguage(LanguageGroup.CSharp));
```
Comment (duncanpMS) [inline src/Integration.UnitTests/LanguageGroupHelperTests.cs]: Typo - "GerLanguage"

File: src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+        [TestMethod]
+        public void QualityProfileBackgroundProcessor_QueueCheckIfUpdateIsRequired_NoProfiles_RequiresUpdate()
+        {
+            // Setup
+            var testSubject = this.GetTestSubject();
+            var project1 = new ProjectMock("validProject1.csproj");
+            project1.SetCSProjectKind();
+            var project2 = new ProjectMock("validProject2.csproj");
+            project2.SetCSProjectKind();
+            this.projectSystem.FilteredProjects = new[] { project1, project2 };
```
Comment (duncanpMS) [inline src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs]: There's a lot of duplication in the setup section across the tests which makes it harder to see what the important differences are. For example, the code to set the filtered projects is repeated numerous times, but it looks like the only important detail is the language of the project. A helper method like _SetFilteredProjects(params languageType)_ would remove the duplication and emphasise the important details.

File: src/Integration.UnitTests/Service/SonarQubeServiceWrapperTests.cs [hunk 11]
```diff
@@ -672,6 +747,17 @@ private static void RegisterProfileExportQueryValidator(TestableSonarQubeService
             });
         }
 
+        private static void RegisterQualityProfileChangeLogValidatorValidator(TestableSonarQubeServiceWrapper testSubject)
```
Comment (duncanpMS) [inline src/Integration.UnitTests/Service/SonarQubeServiceWrapperTests.cs]: Why _XXXValidatorValidator_?

File: src/Integration/Binding/SolutionBindingOperation.cs [hunk 12]
```diff
@@ -114,15 +116,22 @@ public string GetRuleSetFilePath(RuleSetGroup group)
...
             }
 
+            if (profilesMap == null)
+            {
+                throw new ArgumentNullException(nameof(profilesMap));
+            }
+
             this.SolutionFullPath = this.projectSystem.GetCurrentActiveSolution().FullName;
 
+            profilesMap.ToList().ForEach(kv => qualityProfileMap.Add(kv.Key, kv.Value));
```
Comment (duncanpMS) [inline src/Integration/Binding/SolutionBindingOperation.cs]: What's the reason for creating and populating a new List<>? Just so you can use _.ForEach()_ rather than _foreach{}_?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 13]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
                 return serverVM?.Projects.SingleOrDefault(p => ProjectInformation.KeyComparer.Equals(p.Key, projectKey));
             }
         }
+
+        internal /*testing purposes*/ class QualityProfileBackgroundProcessor : IDisposable
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Missing doc comment

File: src/Integration/Binding/BindingWorkflow.cs (hunk 8 above)
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: We verify the C# plugging during connect (before Bind). Either way I didn't change the flow, previously the same flow existed although a bit deeper so if there is no quality profile for a language we stop.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 14]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+            {
+                if (updateAction == null)
+                {
+                    throw new ArgumentNullException(nameof(updateAction));
+                }
+
+                Debug.Assert(this.BackgroundTask == null, "Not expecting this method to be called more than once");
+
+                var projectSystem = this.host.GetService<IProjectSystemHelper>();
+                projectSystem.AssertLocalServiceIsNotNull();
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: The AssertLocalServiceIsNotNull() is used in many places. It would be cleaner to have a method/extension method on the host instead e.g. _GetLocalService<>_ (assuming that local services should never be null)

File: src/Integration.UnitTests/Service/SonarQubeServiceWrapperTests.cs (hunk 11 above)
Comment (eyalantzMS) [inline src/Integration.UnitTests/Service/SonarQubeServiceWrapperTests.cs]: copy-paste error

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 14 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Yes we discussed this with Matthew already, planning to do it as a technical, at the moment being consistent.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 15]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                }
+
+                var solutionBinding = this.host.GetService<ISolutionBindingSerializer>();
+                solutionBinding.AssertLocalServiceIsNotNull();
+
+                var binding = solutionBinding.ReadSolutionBinding();
+                if (binding == null)
+                {
+                    return;
+                }
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Which is likely to be quicker in most cases - fetching filtered solutions or reading the solution binding? Would it be better to check for a solution binding first?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 16]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+
+                if (binding.Profiles == null || binding.Profiles.Count == 0)
+                {
+                    // Old binding, force refresh immediately
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckNoProfiles);
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Presumably there won't be any writes to the general pane after these changes are merged with those for SVS-91?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 17]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                    updateAction();
+                    return;
+                }
+
+                VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheck);
+
+                CancellationToken token = this.TokenSource.Token;
+                this.BackgroundTask = System.Threading.Tasks.Task.Run(() =>
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: This is being run inside VS; why not use the VS threading helpers?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 16 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: That was the reason I raised that bug (we write to output more and more...), so once merged we won't have any writes to the general output window.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 18]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                {
+                    if (this.ProcessQualityProfileChanges(binding, languages, token))
+                    {
+                        this.host.UIDispatcher.BeginInvoke(new Action(()=>
+                        {
+                            // We mustn't update if was cancelled
+                            if(!token.IsCancellationRequested)
+                            {
+                                updateAction();
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: What assumptions is the code making about the thread the _updateAction_ is being run on, if any? If there are any they should be documented/checked.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 15 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: filtered solutions I imagine - everything in memory, binding  - needs to read from a file, this could vary by all sorts of things and skew the averages.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 19]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                            }
+                        }));
+                    }
+                }, token);
+            }
+
+            private bool ProcessQualityProfileChanges(BoundSonarQubeProject binding, IEnumerable<LanguageGroup> projectLanguageGroups, CancellationToken token)
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Documentation: what does the return value indicate - whether an update is required or not?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 18 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: That it needs to be called on the UI thread. Will add an assert.

File: src/Integration/LanguageGroup.cs [hunk 20]
```diff
@@ -1,14 +1,15 @@
...
+// <copyright file="LanguageGroup.cs" company="SonarSource SA and Microsoft Corporation">
 //   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
 //   Licensed under the MIT License. See License.txt in the project root for license information.
 // </copyright>
 //-----------------------------------------------------------------------
 
 namespace SonarLint.VisualStudio.Integration
 {
-    public enum RuleSetGroup
+    public enum LanguageGroup
```
Comment (mjcheetham) [inline src/Integration/LanguageGroup.cs]: Why is this enum needed? It's bijective to the `Language` class members `CSharp`, `VBNET` and `Unknown` (which are singletons and also implement `IEquatable<Language>`).

The logical grouping of files (SonarQube rule set, additional files [not there yet]) is actually by _QualityProfile_.
At the moment SonarQube only supports _one QualityProfile per language_, for any given SQ project. So using `Language` as the grouping is reasonable*.

*unless SQ provides a way to specify different QPs for components within a SQ project, which they don't. At then at that time the grouping becomes the QP key/name/identifer/itself.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 17 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Because this is the simplest to write and test (like in progression), and didn't see anything that those helpers bring and I need.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 21]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                }
+
+                foreach (var keyValue in binding.Profiles)
+                {
+                    LanguageGroup group = keyValue.Key;
+                    ApplicableQualityProfile oldProfileInfo = keyValue.Value;
+
+                    if (!newProfiles.ContainsKey(group))
+                    {
+                        // Not a relevant profile, we should just ignore it.
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Doesn't the check in line 591 means this shouldn't happen? If so, an assertion would be more appropriate.

File: src/Integration/Binding/SolutionBindingOperation.cs [hunk 22]
```diff
@@ -114,15 +116,22 @@ public string GetRuleSetFilePath(RuleSetGroup group)
         #endregion
 
         #region Public API
-        public void Initialize(IEnumerable<Project> projects)
+        public void Initialize(IEnumerable<Project> projects, Dictionary<LanguageGroup, QualityProfile> profilesMap)
```
Comment (mjcheetham) [inline src/Integration/Binding/SolutionBindingOperation.cs]: `IDictionary`

File: src/Integration/LanguageGroup.cs (hunk 20 above)
Comment (eyalantzMS) [inline src/Integration/LanguageGroup.cs]: Since the applied QP can change we can't use 'QP key/name/identifer/itself' as the grouping mechanism since it can change in the middle and we will not be able to handle that gracefully.

Language as group - I considered that but decided that its too risky and unrelated change to do part of this work. So done the simplest thing which is just renaming and map it to language.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 21 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: No, 591 checks that all new profiles are included in the binding profile, 602 checks if the binding profile has a quality profile that was not fetched part of the new profiles since the solution has no project(s) that require that profile.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 23]
```diff
@@ -462,6 +491,190 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                    {
+                        return true;
+                    }
+                }
+
+                VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckQualityProfileIsUpToDate);
+                return false; // Up-to-date
+            }
+
+            private bool IsNewProfileRequriesAnUpdate(QualityProfile newProfileInfo, ApplicableQualityProfile oldProfileInfo)
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Typo - "Requries".
The name is rather cumbersome - does it just mean _HasProfileChanged_?

File: src/Integration/LanguageGroup.cs (hunk 20 above)
Comment (mjcheetham) [inline src/Integration/LanguageGroup.cs]: > it can change in the middle

The Quality Profile can change in the middle of a binding??

> its too risky and unrelated change to do part of this work.

Fair enough. Perhaps we should add a technical ticket?

File: src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs (hunk 10 above)
Comment (eyalantzMS) [inline src/Integration.UnitTests/LocalServices/ErrorListInfoBarController.QualityProfileBackgroundProcessorTests.cs]: fixed

File: src/Integration/LanguageGroup.cs (hunk 20 above)
Comment (eyalantzMS) [inline src/Integration/LanguageGroup.cs]: Yes it can change in the middle of a binding, nothing preventing user to change the profile on the server to something else... Not very likely but possible.
Technical ticket - Sure I'll add one.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 24]
```diff
@@ -462,6 +498,194 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                if (binding == null)
+                {
+                    return;
+                }
+
+                if (binding.Profiles == null || binding.Profiles.Count == 0)
+                {
+                    // Old binding, force refresh immediately
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckNoProfiles);
+                    updateAction(Strings.SonarLintInfoBarOldBindingFile);
```
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Special case: the info bar message will be different in this case (requested by J-M/Blair).

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 25]
```diff
@@ -462,6 +498,194 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
...
+                if (oldProfileInfo.ProfileTimestamp != newProfileInfo.QualityProfileTimestamp)
+                {
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckProfileUpdated);
+                    return true; // The profile was updated
+                }
+
+                return false;
+            }
+
+            private bool TryGetNewProfiles(BoundSonarQubeProject binding, IEnumerable<LanguageGroup> projectLanguageGroups, CancellationToken token, ConnectionInformation connection, out Dictionary<LanguageGroup, QualityProfile> newProfiles)
```
Comment (duncanpMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: "NewProfiles" is implies it is only get the profiles that are new which isn't the case; TryGetLatestProfiles/TryGetServerProfiles would be clearer.

File: src/Integration/Resources/Strings.resx [hunk 26]
```diff
@@ -602,4 +602,42 @@ This is the general format to use when writing errors to output window or when t
...
     <comment>Output window message</comment>
   </data>
+  <data name="SonarQubeOptionalServiceFailed" xml:space="preserve">
+    <value>Warning failed calling '{0}' (http error code={1}). Some of the functionality will be reduced.</value>
+    <comment>Output window message. 
+{0} - service
+{1} - http error code</comment>
+  </data>
+  <data name="SonarLintProfileCheck" xml:space="preserve">
+    <value>SonarLint: Checking if SonarQube Quality profile has changed.</value>
```
Comment (duncanpMS) [inline src/Integration/Resources/Strings.resx]: Presumably the "SonarLint: " prefix is no longer required?

Comment (eyalantzMS) [inline src/Integration/Resources/Strings.resx]: Yes, refresh the page.

File: src/Integration/Resources/Strings.resx [hunk 27]
```diff
@@ -602,4 +602,42 @@ This is the general format to use when writing errors to output window or when t
...
+  <data name="SonarLintProfileCheckProfileUpdated" xml:space="preserve">
+    <value>SonarLint: The Quality Profile for the SonarQube project has changed. Update is required.</value>
+    <comment>Ouput window message</comment>
+  </data>
+  <data name="SonarLintProfileCheckQualityProfileIsUpToDate" xml:space="preserve">
+    <value>SonarLint: Quality Profile is up-to-date.</value>
+    <comment>Ouput window message</comment>
+  </data>
+  <data name="SonarLintProfileCheckSolutionRequiresMoreProfiles" xml:space="preserve">
+    <value>SonarLint: The solution has changed and now requires additional Quality Profiles to be included. Update is required.</value>
```
Comment (duncanpMS) [inline src/Integration/Resources/Strings.resx]: It looks like this message is used when something on the SonarQube server has changed, not when the solution has changed.

Comment (eyalantzMS) [inline src/Integration/Resources/Strings.resx]: That is not correct - it will be shown when the persisted profiles don't have a quality profile for one of more languages that they do need, which could be only when the solution changed.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 28]
```diff
@@ -462,6 +498,194 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
                 return serverVM?.Projects.SingleOrDefault(p => ProjectInformation.KeyComparer.Equals(p.Key, projectKey));
             }
         }
+
+        /// <summary>
+        /// The class is responsible for quality profile related checks in determining whether
+        /// to suggest the user to update his solution with the rule set.
+        /// </summary>
+        internal /*testing purposes*/ class QualityProfileBackgroundProcessor : IDisposable
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Would we not pull this out of `ErrorListInfoBarController` into its own file?

Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: It has no purpose outside of ErrorListInfoBarController.

File: src/Integration/Binding/BindingWorkflow.cs (hunk 8 above)
Comment (duncanpMS) [inline src/Integration/Binding/BindingWorkflow.cs]: That seems like the wrong behaviour and should be documented at least with a TODO/issue.
The code purports to support multiple languages: it should either do the right thing or not pretend to handle anything other than C#.

**Comment by duncanpMS** [comment]
LGTM

File: src/SonarQube/SolutionBinding.sqconfig [hunk 29]
```diff
@@ -1,4 +1,10 @@
...
+    "CSharp": {
```
Comment (mjcheetham) [inline src/SonarQube/SolutionBinding.sqconfig]: Can be not use the server language key? Because this is going to be horrible to try and [correct/account for later](https://jira.sonarsource.com/browse/SVS-94).

Comment (eyalantzMS) [inline src/SonarQube/SolutionBinding.sqconfig]: We should not use SQ values (they could change). So should not be a pain if we'll have a stable display identifier which will have the same value.

File: src/Integration/Binding/BindingWorkflow.cs (hunk 8 above)
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Don't agree here, if you're saying that the plugin is not installed then in terms SQ the project needs to be excluded from analysis. We will fail in this case and which point the user will be able to exclude and retry.

**Comment by mjcheetham** [comment]
LGTM
